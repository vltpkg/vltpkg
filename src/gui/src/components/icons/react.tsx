export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
