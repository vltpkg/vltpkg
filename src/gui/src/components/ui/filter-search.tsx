import { type SavedQuery } from '@/state/types.js'
import { useEffect, useRef, useState } from 'react'
import { Input } from '@/components/ui/input.jsx'
import { Kbd } from '@/components/ui/kbd.jsx'
import { Command, Search } from 'lucide-react'

interface FilterSearchProps<T> {
  items: T[] | undefined
  setFilteredItems: (i: T[]) => void
  placeholder: string
}

const FilterSearch = <T,>({
  items,
  setFilteredItems,
  placeholder,
}: FilterSearchProps<T>) => {
  const [filterText, setFilterText] = useState<string>('')
  const inputRef = useRef<HTMLInputElement>(null)
  const isInitialMount = useRef(true)

  /**
   * Update URL params based on `filterText` after initial load.
   */
  useEffect(() => {
    if (isInitialMount.current) return

    const params = new URLSearchParams(window.location.search)
    const itemKeys = items ? Object.keys(items[0] ?? {}) : []

    if (!filterText.trim()) {
      params.forEach((_, key) => params.delete(key))
      setFilteredItems(items ?? [])
    }

    if (filterText.trim() !== '') {
      params.forEach((_, key) => params.delete(key))
      params.set('filter', filterText)
    }

    if (filterText.split('=')[0] === 'label') {
      params.forEach((_, key) => params.delete(key))
      const label = filterText.split('=')[1]
      if (label) {
        params.set('label', label)
      }
    }

    if (itemKeys.some(key => filterText.includes(key))) {
      params.forEach((_, key) => params.delete(key))
      const [key, value] = filterText.split('=')
      if (key && value) {
        params.set(key, value)
      }
    }

    const newUrl = `${window.location.pathname}?${params.toString()}`
    window.history.replaceState({}, '', newUrl)
  }, [filterText])

  /**
   * Sync URL params with filtered items on initial load.
   */
  useEffect(() => {
    const params = new URLSearchParams(window.location.search)
    const itemKeys = items ? Object.keys(items[0] ?? {}) : []

    const paramKeys = Array.from(params.keys())

    const matchingKey = paramKeys.find(key => itemKeys.includes(key))

    if (matchingKey) {
      setFilterText(`${matchingKey}=${params.get(matchingKey)}`)
    }

    if (params.get('label')) {
      setFilterText(`label=${params.get('label')?.toLowerCase()}`)
    }

    isInitialMount.current = false
  }, [])

  /**
   * Filter items based on `filterText` or URL params.
   */
  useEffect(() => {
    if (!items) {
      setFilteredItems([])
      return
    }

    const params = new URLSearchParams(window.location.search)
    const selectors: { key: string; value: string }[] = []

    for (const [key, value] of params.entries()) {
      selectors.push({ key, value })
    }

    const filteredItems = items.filter(item =>
      selectors.every(selector => {
        if (selector.key === 'filter') {
          const searchValue = selector.value.toLowerCase()
          return Object.values(item as Record<string, unknown>).some(
            val => String(val).toLowerCase().includes(searchValue),
          )
        } else if (selector.key === 'label') {
          const searchValue = selector.value.toLowerCase()
          const labels = (item as SavedQuery).labels ?? []
          return labels.some(label =>
            label.name.toLowerCase().includes(searchValue),
          )
        } else {
          const itemValue = String(
            item[selector.key as keyof T] || '',
          ).toLowerCase()
          const selectorValue = selector.value.toLowerCase()
          return itemValue === selectorValue
        }
      }),
    )

    setFilteredItems(filteredItems)
  }, [items, window.location.search, filterText])

  /**
   * Handle keyboard shortcuts.
   */
  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      switch (true) {
        case (event.metaKey || event.ctrlKey) && event.key === 'k':
          event.preventDefault()
          inputRef.current?.focus()
          break

        case event.key === 'Escape':
          event.preventDefault()
          inputRef.current?.blur()
          break

        default:
          break
      }
    }

    window.addEventListener('keydown', handleKeyDown)

    return () => {
      window.removeEventListener('keydown', handleKeyDown)
    }
  }, [])

  return (
    <div className="relative flex w-[384px] items-center">
      <Search
        size={18}
        className="absolute left-0 ml-3 text-neutral-500"
      />
      <Input
        type="text"
        ref={inputRef}
        className="pl-9 pr-20"
        role="search"
        placeholder={placeholder}
        value={filterText}
        onChange={e => setFilterText(e.target.value)}
      />
      <div className="absolute right-0 flex gap-2">
        <div className="mr-3 hidden items-center gap-1 backdrop-blur-sm md:flex">
          <Kbd>
            <Command size={12} />
          </Kbd>
          <Kbd className="text-sm">k</Kbd>
        </div>
      </div>
    </div>
  )
}

export { FilterSearch }
