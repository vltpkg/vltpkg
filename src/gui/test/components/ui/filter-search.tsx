import React from 'react'
import { vi, describe, it, expect, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { FilterSearch } from '@/components/ui/filter-search.jsx'

vi.mock('@/components/ui/input.jsx', () => ({
  Input: 'gui-input',
}))

vi.mock('@/components/ui/kbd.jsx', () => ({
  Kbd: 'gui-kbd',
}))

vi.mock('lucide-react', () => ({
  Command: 'gui-command-icon',
  Search: 'gui-search-icon',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  cleanup()
})

describe('filter-search', () => {
  it('should render correctly', () => {
    const setStateMock = vi.fn()
    vi.spyOn(React, 'useState').mockReturnValue(['', setStateMock])

    const mockData = [
      {
        id: '01',
        name: 'mock-data-1',
      },
      {
        id: '02',
        name: 'mock-data-2',
      },
      {
        id: '03',
        name: 'mock-data-3',
      },
    ]

    const Container = () => {
      return (
        <FilterSearch
          placeholder="mock-test"
          items={mockData}
          setFilteredItems={setStateMock}
        />
      )
    }

    const { container } = render(<Container />)

    expect(container.innerHTML).toMatchSnapshot()
  })
})
