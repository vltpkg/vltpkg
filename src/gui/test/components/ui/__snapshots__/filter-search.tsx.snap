// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filter-search > should render correctly 1`] = `

<div class="relative flex w-[384px] items-center">
  <gui-search-icon
    size="18"
    classname="absolute left-0 ml-3 text-neutral-500"
  >
  </gui-search-icon>
  <gui-input
    type="text"
    classname="pl-9 pr-20"
    role="search"
    placeholder="mock-test"
    value
  >
  </gui-input>
  <div class="absolute right-0 flex gap-2">
    <div class="mr-3 hidden items-center gap-1 backdrop-blur-sm md:flex">
      <gui-kbd>
        <gui-command-icon size="12">
        </gui-command-icon>
      </gui-kbd>
      <gui-kbd classname="text-sm">
        k
      </gui-kbd>
    </div>
  </div>
</div>

`;
